### PR TITLE
Feature/add feedback stack

### PR DESCRIPTION
**Pull request by dhee-tree**
## Is it running successfully?
This stack should allow you access these routes:
- [feedback](http://localhost:25200/feedback)
- [search](http://localhost:25000/search)

Routes with `dp-frontend-router`
- [feedback](http://localhost:20000/feedback)
- [search](http://localhost:20000/search)

And have a local mail server running:
- [mailhog](http://127.0.0.1:8025/#)

## What to expect
When running you should be able to:
- Access all service with own route and `dp-frontend-router` routes
- Access the `feedback` page, fill in the form and if form is valid, get a thank you message and an email on mailhog
- Access the `search` page
    - Click `Yes`, get a thank you message with no email sent
    - Click `No`, fill in the form and if form is valid, get a thank you message and an email on mailhog

### CODE REVIEW

File: v2/manifests/core-ons/dp-feedback-api.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+      - 28600:28600
+    restart: unless-stopped
+    environment:
+      BIND_ADDR:                          ":28600"
+      GRACEFUL_SHUTDOWN_TIMEOUT:          ${GRACEFUL_SHUTDOWN_TIMEOUT:-5s}
+      HEALTHCHECK_INTERVAL:               ${HEALTHCHECK_INTERVAL:-30s}
+      HEALTHCHECK_CRITICAL_TIMEOUT:       ${HEALTHCHECK_CRITICAL_TIMEOUT:-90s}
+      MAIL_HOST:                          ${MAIL_HOST:-mailhog}
+      MAIL_ENCRYPTION:                    ${MAIL_ENCRYPTION:-true}
+      API_ROUTER_URL:                     ${API_ROUTER_URL:-http://dp-api-router:23200}
```
Comment (lindenmckenzie) [inline v2/manifests/core-ons/dp-feedback-api.yml:27]: Not sure you need the `API_ROUTER_URL`?

File: v2/manifests/core-ons/dp-frontend-feedback-controller.yml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    ports:
+      - 25200:25200
+    restart: unless-stopped
+    environment:
+      BIND_ADDR:                          ":25200"
+      MAIL_HOST:                          ${MAIL_HOST:-mailhog}
+      FEEDBACK_API_URL:                   ${FEEDBACK_API_URL:-http://dp-feedback-api:28600}
+      API_ROUTER_URL:                     ${API_ROUTER_URL:-http://dp-api-router:23200}
+      ENABLE_FEEDBACK_API:                ${ENABLE_FEEDBACK_API:-false}
+      MAIL_ENCRYPTION:                    ${MAIL_ENCRYPTION:-true}
```
Comment (lindenmckenzie) [inline v2/manifests/core-ons/dp-frontend-feedback-controller.yml:26]: Could we A-Z these plase?

File: v2/manifests/core-ons/dp-frontend-router.yml [hunk 3]
```diff
@@ -29,6 +29,9 @@ services:
       RELEASE_CALENDAR_CONTROLLER_URL: ${RELEASE_CALENDAR_CONTROLLER_URL:-http://dp-frontend-release-calendar:27700}
       SEARCH_ROUTES_ENABLED:           "true"
       SEARCH_CONTROLLER_URL:           ${SEARCH_CONTROLLER_URL:-http://dp-frontend-search-controller:25000}
+      FEEDBACK_CONTROLLER_URL:         ${FEEDBACK_CONTROLLER_URL:-http://dp-frontend-feedback-controller:25200}
+      FEEDBACK_ENABLED:                ${FEEDBACK_ENABLED:-false}
```
Comment (lindenmckenzie) [inline v2/manifests/core-ons/dp-frontend-router.yml:33]: This flag isn't actually doing anything.

File: v2/manifests/core-ons/dp-frontend-search-controller.yml [hunk 4]
```diff
@@ -18,10 +18,12 @@ services:
       - 25000:25000
     restart: unless-stopped
     environment:
-      BIND_ADDR:          ":25000"
-      API_ROUTER_URL:     ${API_ROUTER_URL:-http://dp-api-router:23200}/v1
-      IS_PUBLISHING:      ${IS_PUBLISHING:-true}
-      SERVICE_AUTH_TOKEN: ${SERVICE_AUTH_TOKEN:?please define a valid SERVICE_AUTH_TOKEN in your local system}
+      BIND_ADDR:                ":25000"
+      API_ROUTER_URL:           ${API_ROUTER_URL:-http://dp-api-router:23200}/v1
```
Comment (lindenmckenzie) [inline v2/manifests/core-ons/dp-frontend-search-controller.yml:22]:
```suggestion
      API_ROUTER_URL:           ${API_ROUTER_URL:-http://dp-api-router:23200/v1}
```

File: v2/manifests/core-ons/dp-frontend-search-controller.yml [hunk 5]
```diff
@@ -18,10 +18,12 @@ services:
...
+      IS_PUBLISHING:            ${IS_PUBLISHING:-true}
```
Comment (lindenmckenzie) [inline v2/manifests/core-ons/dp-frontend-search-controller.yml:23]: Not sure why we're defaulting to true here, but equally not your change.

File: v2/stacks/feedback/README.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+This stack deploys the necessary services and dependencies for feedback service including the feedback api, upstream services and local mail server.
+
+The stack uses:
+- dp-frontend-feedback-controller
+- dp-frontend-search-controller
+- dp-frontend-router
+- dp-feedback-api
+- dp-api-router
+- dp-design-system
+- zebedee-reader
```
Comment (lindenmckenzie) [inline v2/stacks/feedback/README.md:12]: why do we need zebedee-reader?

File: v2/stacks/feedback/README.md [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+- elasticsearch
```
Comment (lindenmckenzie) [inline v2/stacks/feedback/README.md:13]: why do we need elasticsearch?

File: v2/stacks/feedback/README.md [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+- mailhog
+
+## Dependencies
+
+This stack relies on having a local zebedee content store setup correctly and the correct path set to $zebedee_root
+
+See the [zebedee getting started guide](https://github.com/ONSdigital/zebedee#getting-started) for more information.
```
Comment (lindenmckenzie) [inline v2/stacks/feedback/README.md:20]: Why do we need zebedee?

File: v2/stacks/feedback/README.md [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+## Is it running successfully?
```
Comment (lindenmckenzie) [inline v2/stacks/feedback/README.md:22]: I'd recommend turning a markdown linter on for this file.

File: v2/stacks/feedback/README.md [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+You should:
+- rename the `local.env.tmpl` => `local.env`
+- This would override these configs for the following services
+    - ENABLE_FEEDBACK_API to `true`
+        - `dp-api-router`
+        - `dp-frontend-feedback-controller`
+        - `dp-frontend-search-controller`
+    - FEEDBACK_ENABLED to `true`
+        - `dp-frontend-router`
```
Comment (lindenmckenzie) [inline v2/stacks/feedback/README.md:45]: I'm not sure this flag does anything.

File: v2/stacks/feedback/README.md [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+You can change it back to `false` to turn off the feature flag for `dp-feedback-api`
+
+## What to expect
+When running you should be able to:
+- Access the feedback and search service with own route and `dp-frontend-router` routes
```
Comment (lindenmckenzie) [inline v2/stacks/feedback/README.md:51]:
```suggestion
- Access the feedback and search service with their own routes and `dp-frontend-router` routes
```

File: v2/stacks/feedback/README.md [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+- Access the `feedback` page, fill in the form and if form is valid, get a thank you message and an email on mailhog
```
Comment (lindenmckenzie) [inline v2/stacks/feedback/README.md:52]:
```suggestion
- Access the `feedback` page, fill in the form and if the form is valid, get a thank you message and an email on mailhog
```

File: v2/stacks/feedback/README.md (hunk 6 above)
Comment (lindenmckenzie) [reply v2/stacks/feedback/README.md:12]: do we need it for the dp-api-router? that seems a pain.

File: v2/stacks/feedback/core-ons.yml [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+        condition: service_healthy
+      dp-frontend-router:
+        condition: service_healthy
+  dp-feedback-api:
+    extends:
+      file: ${PATH_MANIFESTS}/core-ons/dp-feedback-api.yml
+      service: dp-feedback-api
+    depends_on:
+      mailhog:
+        condition: service_healthy
```
Comment (lindenmckenzie) [inline v2/stacks/feedback/core-ons.yml:26]: I think this file would be better split into:
- supporting services (routers, design-system etc)
- feedback service (api, controller)
- upstream services (search controller)

**Review by lindenmckenzie — CHANGES_REQUESTED** [review]
A few bits, and possibly some discussion.

File: v2/manifests/core-ons/dp-frontend-feedback-controller.yml [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+      - "25200"
+    ports:
+      - 25200:25200
+    restart: unless-stopped
+    environment:
+      API_ROUTER_URL:                     ${API_ROUTER_URL:-http://dp-api-router:23200}
+      BIND_ADDR:                          ":25200"
+      ENABLE_FEEDBACK_API:                ${ENABLE_FEEDBACK_API:-false}
+      FEEDBACK_API_URL:                   ${FEEDBACK_API_URL:-http://dp-feedback-api:28600}
+      MAIL_ENCRYPTION:                    ${MAIL_ENCRYPTION:-true}
```
Comment (lindenmckenzie) [inline v2/manifests/core-ons/dp-frontend-feedback-controller.yml:25]: shouldn't this be false?

File: v2/stacks/feedback/README.md [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+You should:
+
+- rename the `local.env.tmpl` => `local.env`
+- This would override these configs for the following services
+  - ENABLE_FEEDBACK_API to `true`
+        - `dp-api-router`
+        - `dp-frontend-feedback-controller`
+        - `dp-frontend-search-controller`
+
+You can change it back to `false` to turn off the feature flag for `dp-feedback-api`
```
Comment (lindenmckenzie) [inline v2/stacks/feedback/README.md:45]:
```suggestion
You can change it back to `false` to turn off the feature flag for `dp-feedback-api`.
```

File: v2/stacks/feedback/README.md [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+
+## What to expect
+
+When running you should be able to:
+
+- Access the feedback and search service with their own routes and `dp-frontend-router` routes
+- Access the `feedback` page, fill in the form and if the form is valid, get a thank you message and an email on mailhog
+- Access the `search` page
```
Comment (lindenmckenzie) [inline v2/stacks/feedback/README.md:53]:
```suggestion
- Access the `search` page and interact with the feedback footer:
```

File: v2/stacks/feedback/default.env [hunk 17]
```diff
@@ -0,0 +1,15 @@
+# -- Global variable overrides --
+IS_PUBLISHING="false"
+MAIL_ENCRYPTION=${MAIL_ENCRYPTION:-false}
```
Comment (lindenmckenzie) [inline v2/stacks/feedback/default.env:3]: Doesn't need to be here - put this in the manifest instead.

File: v2/stacks/feedback/default.env [hunk 18]
```diff
@@ -0,0 +1,15 @@
...
+ZEBEDEE_URL=http://http-echo:5678
```
Comment (lindenmckenzie) [inline v2/stacks/feedback/default.env:4]: Add a comment here to explain we're stubbing the health check.

File: v2/stacks/feedback/deps.yml [hunk 19]
```diff
@@ -0,0 +1,9 @@
+services:
+  mailhog:
+    extends:
+      file: ${PATH_MANIFESTS}/deps/mailhog.yml
+      service: mailhog
+  http-echo:
+    extends:
+      file: ${PATH_MANIFESTS}/deps/http-echo.yml
+      service: http-echo
```
Comment (lindenmckenzie) [inline v2/stacks/feedback/deps.yml:9]: Could put comment here to explain this is not part of the service, but is to stub out health checks.

File: v2/stacks/feedback/feedback-service.yml [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+services:
+  dp-frontend-feedback-controller:
+    extends:
+      file: ${PATH_MANIFESTS}/core-ons/dp-frontend-feedback-controller.yml
+      service: dp-frontend-feedback-controller
+    depends_on:
+      dp-design-system:
+        condition: service_healthy
+      dp-frontend-router:
+        condition: service_healthy
```
Comment (lindenmckenzie) [inline v2/stacks/feedback/feedback-service.yml:14]: I think this could go.

File: v2/stacks/feedback/supporting-services.yml [hunk 21]
```diff
@@ -0,0 +1,20 @@
+# Supporting Services
+
+# Contains all the supporting services for the feedback stack
+
+services:
+  dp-api-router:
+    extends:
+      file: ${PATH_MANIFESTS}/core-ons/dp-api-router.yml
+      service: dp-api-router
```
Comment (lindenmckenzie) [inline v2/stacks/feedback/supporting-services.yml:9]: depends on http-echo.

File: v2/stacks/feedback/supporting-services.yml [hunk 22]
```diff
@@ -0,0 +1,20 @@
...
+  dp-frontend-router:
+    extends:
+      file: ${PATH_MANIFESTS}/core-ons/dp-frontend-router.yml
+      service: dp-frontend-router
+    depends_on:
+      dp-api-router:
+        condition: service_healthy
```
Comment (lindenmckenzie) [inline v2/stacks/feedback/supporting-services.yml:16]: I don't think this is a dependency.

File: v2/stacks/feedback/upstream-services.yml [hunk 23]
```diff
@@ -0,0 +1,14 @@
+# Upstream Services
+
+# Contains all the upstream services for the feedback stack
```
Comment (lindenmckenzie) [inline v2/stacks/feedback/upstream-services.yml:3]: This isn't technically true. 

We could put a list here of other services you could put in if you want. 
```suggestion
# Contains examples of upstream services for the feedback stack
# Other services which use the feedback footer include:
# - dp-frontend-homepage-controller 
```
...etc.

File: v2/stacks/feedback/upstream-services.yml [hunk 24]
```diff
@@ -0,0 +1,14 @@
...
+services:
+  dp-frontend-search-controller:
+    extends:
+      file: ${PATH_MANIFESTS}/core-ons/dp-frontend-search-controller.yml
+      service: dp-frontend-search-controller
+    depends_on:
+      dp-design-system:
+        condition: service_healthy
+      dp-frontend-router:
+        condition: service_healthy
```
Comment (lindenmckenzie) [inline v2/stacks/feedback/upstream-services.yml:14]: Take this out.

File: v2/manifests/core-ons/dp-feedback-api.yml [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+      - "28600"
+    ports:
+      - 28600:28600
+    restart: unless-stopped
+    environment:
+      BIND_ADDR:                          ":28600"
+      GRACEFUL_SHUTDOWN_TIMEOUT:          ${GRACEFUL_SHUTDOWN_TIMEOUT:-5s}
+      HEALTHCHECK_CRITICAL_TIMEOUT:       ${HEALTHCHECK_CRITICAL_TIMEOUT:-90s}
+      HEALTHCHECK_INTERVAL:               ${HEALTHCHECK_INTERVAL:-30s}
+      MAIL_ENCRYPTION:                    ${MAIL_ENCRYPTION:-true}
```
[the remark this answers is not in the document]
Comment (dhee-tree) [reply v2/manifests/core-ons/dp-feedback-api.yml]: Done

File: v2/stacks/feedback/README.md (hunk 15 above)
Comment (dhee-tree) [reply v2/stacks/feedback/README.md]: Done

File: v2/stacks/feedback/README.md (hunk 16 above)
Comment (dhee-tree) [reply v2/stacks/feedback/README.md]: Done

File: v2/stacks/feedback/default.env (hunk 17 above)
Comment (dhee-tree) [reply v2/stacks/feedback/default.env]: Done

File: v2/stacks/feedback/default.env (hunk 18 above)
Comment (dhee-tree) [reply v2/stacks/feedback/default.env]: Done

File: v2/stacks/feedback/deps.yml (hunk 19 above)
Comment (dhee-tree) [reply v2/stacks/feedback/deps.yml]: Done

File: v2/stacks/feedback/feedback-service.yml (hunk 20 above)
Comment (dhee-tree) [reply v2/stacks/feedback/feedback-service.yml]: Done

File: v2/stacks/feedback/supporting-services.yml (hunk 21 above)
Comment (dhee-tree) [reply v2/stacks/feedback/supporting-services.yml:9]: Done

File: v2/stacks/feedback/supporting-services.yml (hunk 22 above)
Comment (dhee-tree) [reply v2/stacks/feedback/supporting-services.yml]: Done

File: v2/stacks/feedback/upstream-services.yml (hunk 23 above)
Comment (dhee-tree) [reply v2/stacks/feedback/upstream-services.yml]: Done

File: v2/stacks/feedback/upstream-services.yml (hunk 24 above)
Comment (dhee-tree) [reply v2/stacks/feedback/upstream-services.yml]: Done

File: v2/manifests/core-ons/dp-api-router.yml [hunk 26]
```diff
@@ -43,6 +43,9 @@ services:
       FILES_API_URL:               ${FILES_API_URL:-http://dp-files-api:26900}
       IDENTITY_API_URL:            ${IDENTITY_API_URL:-http://dp-identity-api:25600}
       DOWNLOAD_SERVICE_URL:        ${DOWNLOAD_SERVICE_URL:-http://dp-download-service:23600}
+      FEEDBACK_CONTROLLER_URL:     ${FEEDBACK_CONTROLLER_URL:-http://dp-frontend-feedback-controller:25200}
```
Comment (lindenmckenzie) [inline v2/manifests/core-ons/dp-api-router.yml:46]: Why's this here?

File: v2/manifests/core-ons/dp-frontend-router.yml [hunk 27]
```diff
@@ -18,17 +18,19 @@ services:
...
     restart: unless-stopped
     environment:
-      HEALTHCHECK_INTERVAL:            ${HEALTHCHECK_INTERVAL:-30s}
-      HOMEPAGE_CONTROLLER_URL:         ${HOMEPAGE_CONTROLLER_URL:-http://dp-frontend-homepage-controller:24400}
+      API_ROUTER_URL:                  ${API_ROUTER_URL:-http://dp-api-router:23200}/v1
       BABBAGE_URL:                     ${BABBAGE_URL:-http://babbage:8080}
       DATASET_CONTROLLER_URL:          ${DATASET_CONTROLLER_URL:-http://dp-frontend-dataset-controller:20200}
-      API_ROUTER_URL:                  ${API_ROUTER_URL:-http://dp-api-router:23200}/v1
-      FILTER_FLEX_ROUTES_ENABLED:      ${FILTER_FLEX_ROUTES_ENABLED:-false}
+      FEEDBACK_API_URL:                ${FEEDBACK_API_URL:-http://dp-feedback-api:28600}
```
Comment (lindenmckenzie) [inline v2/manifests/core-ons/dp-frontend-router.yml:24]: Why's this here?

File: v2/manifests/core-ons/dp-frontend-search-controller.yml [hunk 28]
```diff
@@ -18,10 +18,12 @@ services:
...
     restart: unless-stopped
     environment:
-      BIND_ADDR:          ":25000"
-      API_ROUTER_URL:     ${API_ROUTER_URL:-http://dp-api-router:23200}/v1
-      IS_PUBLISHING:      ${IS_PUBLISHING:-true}
-      SERVICE_AUTH_TOKEN: ${SERVICE_AUTH_TOKEN:?please define a valid SERVICE_AUTH_TOKEN in your local system}
+      API_ROUTER_URL:           ${API_ROUTER_URL:-http://dp-api-router:23200/v1}
+      BIND_ADDR:                ":25000"
+      ENABLE_FEEDBACK_API:      ${ENABLE_FEEDBACK_API:-false}
+      FEEDBACK_CONTROLLER_URL:  ${FEEDBACK_CONTROLLER_URL:-http://dp-frontend-feedback-controller:25200}
```
Comment (lindenmckenzie) [inline v2/manifests/core-ons/dp-frontend-search-controller.yml:24]: Think this should be `FEEDBACK_API_URL`

File: v2/manifests/core-ons/dp-api-router.yml (hunk 26 above)
Comment (dhee-tree) [reply v2/manifests/core-ons/dp-api-router.yml]: Done

File: v2/manifests/core-ons/dp-frontend-router.yml (hunk 27 above)
Comment (dhee-tree) [reply v2/manifests/core-ons/dp-frontend-router.yml]: Done

File: v2/manifests/core-ons/dp-frontend-search-controller.yml (hunk 28 above)
Comment (dhee-tree) [reply v2/manifests/core-ons/dp-frontend-search-controller.yml]: Done

**Review by lindenmckenzie — APPROVED** [review]

**Pull request merged**